### PR TITLE
Fix SDK Docker build scripts, update Emscripten

### PR DESCRIPTION
**Pull request by krinkinmu**
The way the script is currently it does not work (anymore). I'm aware of two issues with the current setup:

1. emsdk underneath installs and uses Node JS and the version of Node JS is not pinned by the script and isn't tied to the version of the SDK used.

   As a result, emsdk moved to a newer version of Node JS than the one that was probably used when the scripts were created. This new version (it seems, the version currently being used is 18.20.3) requires a relatively fresh version of glibc which just isn't available in Ubuntu Bionic used as a base image. That results in obscure errors from CMake (see https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/170)
2. for whatever reason, the version of the protobuf static libraries currently in the repo, don't seem to work. I don't really know how the issue happened, but there were at least 2 users that faced the problem (see https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/161).

To resolve the first issue, we have a bunch of options:

1. Fix the version of emsdk
2. Fix the version of Node JS
3. Update the version of glibc

I figured I can combine options 1 and 3 together for the following reasons:

1. Fixing the version of emsdk fixes the problem
2. The problem arised from the fact that the versions of software used in the build script are a bit old, so an update might be in order, even though we have other solutions to the problem.

> NOTE: It's my understanding that fixing emsdk version should also
> pin Node JS version, so if we deploy option 1, option 2 seem
> redundant.

For the second problem, I think there is only one ultimate solution
- not store binary artifacts in the repository and instead build them from the sources in the repo.

It seems that in the past there was a concern that building protobuf libraries takes a long time - it's still certainly the case. However, I think we can compensate for that in two ways:

1. Drop WAVM - it does not seem like WAVM is still needed (the project itself appears to be dead and hasn't had any updates for at least 2 years), but also one of the comments in https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/pull/158, that removed the WAVM from the docs, also suggests that three doesn't seem to be a good reason to keep WAVM in the SDK build script.
2. Take advantage of potential hardware threads in the system when calling make - most laptops or servers these days have multiple cores, so if we have to build protobuf libraries twice, we can speed it up by using more cores.

With all those changes made, the build time adds up to something like 32 minutes on my laptop, compared to the 48m without building protobuf libraries from sources (and without adding -j option to make). So I think the additional time spent on building protobuf library is compensated by other changes.

### CODE REVIEW

**Comment by martijneken** [comment]
Thanks for the PR! Few thoughts from my end:

* WAVM removal SGTM (see also https://github.com/envoyproxy/envoy/pull/32872)
* The upgrade to emsdk 3.1.67 is very welcome. I think we should update the Bazel toolchain at the same time, see also https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/pull/157. I struggled with this due to Envoy incompatibility (https://github.com/envoyproxy/envoy/pull/32432#issuecomment-1969936133), but I can take another look since it has been a while.
* Eek, I had no idea we had binary files in this repo (libprotobuf). In light of the XZ Backdoor this makes me very uncomfortable. @PiotrSikora @mpwarres WDYT, can we build from source instead?

**Comment by krinkinmu** [comment]
@martijneken I think I have a working change that updates Bazel build files to the newer version of emsdk, but I hit a bit of an issue.

I can see that currently Bazel configuration patches emsdk. Digging through the history, it seems like the patch was added to work around some build issues (see https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/149). Reading through the issue it seems like it manifested when building some proxy-wasm-cpp-host tests (https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/149#issuecomment-1523375750).

With that in mind, I tried to reproduce the issue, so that I can generate and verify the refreshed version of the patch. However, I don't seem to be able to reproduce the issue. It's a long shot, but by any chance do you remeber under what circumstances https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/149 manifested?

NOTE: I will still try to reproduce it myself, I'm just throwing this question in in case you might remember the context.

**Comment by krinkinmu** [comment]
@martijneken I added emsdk update for Bazel build scripts as well now. PTAL.

One thing that I want to call out. I tried to reproduce the problem reported in https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/149 locally to test a new emsdk patch, however, I wasn't able to do that.

I don't think that the patch is needed since https://github.com/emscripten-core/emsdk/pull/1045 which removes the npm modules as a dependency for the toolchain (among other things) and that should uktimately achieve the same effect. That being said, given that I couldn't reproduce the original issue, I wasn't able to explicitly confirm that.

**Comment by krinkinmu** [comment]
@martijneken if you note somewhere what tests you're running manually, I will make sure to run them myself going forward (and include the results) and maybe automate them as well.

**Comment by martijneken** [comment]
> what tests you're running manually

More or less trying to replicate the CI builds and tests in 1 or 2 configurations:
- for CppHost I did: `bazelisk test --config=clang --define engine=null //test/...` and `engine=v8` for "real" wasm
- for Envoy I did: `bazelisk build -c opt //source/exe:envoy-static` and `bazelisk test -c opt //test/extensions/...`

Envoy and v8 builds take a long time, still running. Will send the Envoy PR once this succeeds.

**Comment by martijneken** [comment]
All looks good locally. CI tests:

- [PASSED] https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/422
- [waiting] https://github.com/envoyproxy/envoy/pull/36526

**Review by martijneken — CHANGES_REQUESTED** [review]
Can you update: https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/main/docs/building.md

- update the emsdk version
- remove the libprotobuf.a section

I will send a follow-up PR to add full support for Abseil and RE2.

**Comment by martijneken** [comment]
LGTM after docs fix.

I have a follow-up PR to get full Abseil/RE2 support \o/ https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/pull/173

File: sdk_container.sh
Comment (martijneken) [inline sdk_container.sh:77]: Since we're copying these files to /sdk/ now, I suspect we need to update the various Makefile references? (3 files to look at)

Comment (krinkinmu) [reply sdk_container.sh:77]: We copied files to sdk before, the difference now is that copying happens inside sdk_container.sh and not in Dockerfile-sdk. That's because sdk_container.sh is now responsible for building it and we stop using prebuilt binaries. Everything else remains the same.

FWIW, I confirmed that Makefile.base and Makefile.base_lite work with the example from the docs (https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/main/docs/building.md), so I don't think that it's broken, at least not in an obvious way.

**Comment by krinkinmu** [comment]
> Can you update: https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/main/docs/building.md
> 
> * update the emsdk version
> * remove the libprotobuf.a section
> 
> I will send a follow-up PR to add full support for Abseil and RE2.

I updated the docs with the new emsdk version. I didn't completely remove protobuf section though. I might be missing something, but the way I understand the docs it tries to explain how to prepare all the dependencies without using Docker container or relying on Bazel. So I updated the wording in the docs, but kept the section about protobuf libraries in it.

File: sdk_container.sh
Comment (martijneken) [reply sdk_container.sh:77]: Great! I missed that it was a noop. Do we still need this Makefile target?
https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/54167e2ba495ba603561f4560e527bda22fc9816/Makefile#L14-L25

Either way we can remove the wasm-patches bit.

**Review by martijneken — APPROVED** [review]

**Review by mpwarres — APPROVED** [review]
This looks great--thanks for the contribution!

**Comment by martijneken** [comment]
@krinkinmu I don't recall how the repo permissions are set. If you need us to merge this PR, give a shout when ready.

File: sdk_container.sh
Comment (krinkinmu) [reply sdk_container.sh:77]: A good catch, I missed this one! I added a commit that fixes this Makefile as well, thank you!

**Comment by krinkinmu** [comment]
@martijneken it seems that I cannot merge it myself. Changing protection rules is welcome (to allow merging approved PRs), but in the meantime would you mind pushing a button for me?

**Comment by martijneken** [comment]
ACK, I will look at the repo rules. Thanks for your PR, this is awesome!

**Comment by martijneken** [comment]
Something we forgot about in review: there have been some changes to Emscripten memory management:
- default stack size is now 64 KiB (STACK_SIZE)
- default heap size is now 16 MiB (INITIAL_HEAP) and does not grow by default (ALLOW_MEMORY_GROWTH)

We will send a follow-up PR to tweak these values a bit for server environments.

**Comment by mpwarres** [comment]
> We will send a follow-up PR to tweak these values a bit for server environments.

Sent #174, PTAL

**Comment by krinkinmu** [comment]
@martijneken sorry my bad, I didn't check it - I can send a follow up PR shortly.

I've attempted to trace what the previous values were and here is what I found so far:

1. STACK_SIZE - 5MiB (changed in https://github.com/emscripten-core/emscripten/commit/157fcd4650161d9e0261966e2d3b529d62e1babc)
2. INITIAL_HEAP - this seem to be a relatively new option (introduced in https://github.com/emscripten-core/emscripten/commit/22a1c046261f6b64efbc40fb837b51efe91f2aa9) and before that INITIAL_MEMORY was used and was 16MiB for a long time
3. ALLOW_MEMORY_GROWTH seem to have been disabled since the time it was introduced.

It looks like STACK_SIZE is the only thing that changed, unless I'm missing something.
It does seem like allowing memory to grow is a more sensible default for the servers.

As for INITIAL_HEAP, I'm not certain what value it should be and how harsh the penalty for setting it too low and growing the heap later. Would it make sense to leave it as-is for now, assuming my understanding that it didn't change is correct?

It looks like these options are not currently configurable from proxy_wasm_cc_binary rule. I can expose those as configuration options along the way?

**Comment by PiotrSikora** [comment]
Regarding removed patches for `node_modules` (see: https://github.com/envoyproxy/envoy/pull/21487), the reason for them was two-fold:
1. Downloading them completely broke Envoy CI at the time due to flaky origin for those modules.
2. Emscripten SDK tarball already contains those modules, so reinstalling them is completely unnecessarily.

Separately, you should either cc or add @kyessenov to `CODEOWNERS` as a gatekeepr (assuming it's fine with him), since he knows a lot of history behind those changes, and because Istio is probably still the biggest consumer of this SDK.

**Pull request merged**
